### ISSUE TITLE
Case insensitive isearch

### ISSUE THREAD

**Issue by asmeurer** [benma/visual-regexp-steroids.el#15]
I have enabled case insensitivity in the vr/default-regexp-modifiers, but vr/isearch-forward is still case sensitive. Ideally vr/isearch-forward would be insensitive by default anyway (or use something similar to the [default isearch behavior](https://www.gnu.org/software/emacs/manual/html_node/emacs/Search-Case.html)).

**Comment by benma** [comment]
In the search prompt of `vr/replace`, you can toggle the modifiers with `C-c i`, `C-c m`, etc., and they generate the proper prefix, i.e. `(?im)`. 

Adding the same feature to isearch would be quite a hassle, I believe.

You can always manually type `(?i)` before your search string to enable case insensitivity, as well as use any of the other modifiers. Is that enough as a workaround?

**Comment by asmeurer** [comment]
I'd like it to always be insensitive. But maybe I can write up some code that always adds `(?i)` to the search.

**Comment by asmeurer** [comment]
Note that this is even if I have ` '(vr/default-regexp-modifiers (quote (:I t :M t :S nil :U t)))` set in my customize configuration. It does appear that `C-c i` works in `vr/isearch-forward`, but the defaults from `vr/default-regexp-modifiers` are not respected.

**Comment by asmeurer** [comment]
The advice from https://github.com/benma/visual-regexp-steroids.el/issues/12#issuecomment-135772132 works for me as a workaround, but I don't see why this shouldn't work by default (assuming the isearch settings are configured to ignore case), and I especially don't see why it shouldn't respect the vr settings.

**Comment by benma** [comment]
> It does appear that `C-c i` works in `vr/isearch-forward`

I don't think it does. It is defined in `vr/minibuffer-regexp-keymap`, which is not used during isearch.

As mentioned [here](https://github.com/benma/visual-regexp-steroids.el/issues/12#issuecomment-135755887), I obeyed the isearch case-fold-search variable in the past, but removed it, unfortunately without writing down the reason in the commit message :( I vaguely remember that there was some trouble with that. However, I don't see a problem in adding this feature if it doesn't break anything.

However, I don't see a reason not to implement this feature if it behaves okay.

So the task consists of two points:

1. Add the key bindings to manipulate the modifiers to isearch.  
1. The vr modifiers have only been built for the Python engine and are visible in the prompt when you do the search/replace. They should also be visible in the isearch prompt I guess, because it is an input helper, producing a literal (?modifiers) prefix before the search string. If they are invisible, the search behaves in unexpected ways.

Then there is the question on whether to use the regular isearch key map to manipulate the `case-fold-search`, and whether to obey it in the first place, and how to handle it with the other engines (pcre2el).

**Comment by asmeurer** [comment]
`C-c i` is definitely working for me. I'm on the latest master for visual-regexp.el and visual-regexp-steroids.el. Maybe I have something in my [configuration](https://github.com/asmeurer/dotfiles/blob/master/.emacs) that is making it work.

**Comment by benma** [comment]
Can you check with `emacs -q` and loading only the vr packages? Hitting `C-c i` during vr/isearch-forward only quits the search for me, and I don't see any code in `vr` enabling this in isearch.

Do `C-c m` etc. also work for you, then?

**Comment by asmeurer** [comment]
I get the same behavior with emacs -q. I've got a few isearch customizations (including isearch+), so those may be interacting somehow. With my emacs configuration, `C-c` starts like it's going to cancel the search, but when I type `i` it adds the `(?i)` to the beginning. `C-c m` does not work for me.

**Comment by benma** [comment]
This is very bizarre. In vr/isearch-forward, there is no code to prepend the `(?...)` prefix, let alone make that visible in the prompt. 

> I've got a few isearch customizations (including isearch+), so those may be interacting somehow.

What do you mean? With `emacs -q`, nothing should be loaded that could interfere.

**Comment by asmeurer** [comment]
No I mean in my .emacs I do. With emacs -q it doesn't happen (sorry, I meant to type "I get the same behavior **as you**" above).

**Comment by scmanjarrez** [comment]
I've modified a bit the code provided in https://github.com/benma/visual-regexp-steroids.el/issues/12#issuecomment-135769101 to make it toggleable (sorry, I've little knowledge of elisp):
```elisp
;; Togle case sensitiveness for vr/isearch
(defun toggle-vr-case-insensitive ()
  "Toggle case-insensitive search for visual-regexp."
  (interactive)
  (setq advice-list (list))
  (advice-mapc (lambda (advice _props) (push advice advice-list)) 'vr--isearch)
  (message "advices %d" (length advice-list))
  (if (= (length advice-list) 0)
      (message "Enabling case-insensitive search"
               (defadvice vr--isearch (around add-case-insensitive (forward string &optional bound noerror count) activate)
                 (setq string (concat "(?i)" string))
                 ad-do-it))
    (message "Disabling case-insensitive search"
             (advice-mapc (lambda (advice _props) (advice-remove 'vr--isearch advice)) 'vr--isearch)))
  )
(global-set-key (kbd "C-c t") 'toggle-vr-case-insensitive)
```
